File: Baekjoon/Java/q11003_최솟값찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;
import java.util.Deque;

public class Main{
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int L = Integer.parseInt(st.nextToken());
        int[] arr = new int[N];

        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < N; i++){
            arr[i] = Integer.parseInt(st.nextToken());
        }

        Deque<int[]> deq = new ArrayDeque<>();
        for(int i = 0; i < N; i++){
            if(!deq.isEmpty() && deq.peekFirst()[1] <= i - L){
                deq.pollFirst();
            }

            while(!deq.isEmpty() && deq.peekLast()[0] > arr[i]){
                deq.pollLast();
            }

            deq.add(new int[]{arr[i], i});
            sb.append(deq.peekFirst()[0]).append(" ");
        }

        System.out.println(sb.toString());
    }
}

File: Baekjoon/Python/q11003_최솟값찾기.py
import sys
from collections import deque


N, L = map(int, sys.stdin.readline().split())
arr = list(map(int, sys.stdin.readline().strip().split()))
deq = deque()
answer = []

for i in range(N):

    if deq and deq[0][1] <= i - L:
        deq.popleft()

    while deq and deq[-1][0] > arr[i]:
        deq.pop()

    deq.append([arr[i], i])
    answer.append(deq[0][0])

print(*answer)